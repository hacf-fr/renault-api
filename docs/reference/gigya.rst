Gigya low-level authentication API Reference
============================================

.. contents::
    :local:
    :backlinks: none


Core
----

.. automodule:: renault_api.gigya
   :members:

Data models
-----------

.. automodule:: renault_api.gigya.models
   :members:
