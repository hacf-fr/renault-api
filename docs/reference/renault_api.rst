Renault high-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Session
---------------

.. automodule:: renault_api.renault_session
   :members:

Client
--------------

.. automodule:: renault_api.renault_client
   :members:

Account
---------------

.. automodule:: renault_api.renault_account
   :members:

Vehicle
---------------

.. automodule:: renault_api.renault_vehicle
   :members:
