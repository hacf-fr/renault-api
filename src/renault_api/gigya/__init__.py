"""Gigya API."""
import logging
from typing import Any
from typing import cast
from typing import Dict

import aiohttp
from marshmallow.schema import Schema

from . import models
from . import schemas

GIGYA_JWT = "gigya_jwt"
GIGYA_LOGIN_TOKEN = "gigya_login_token"
GIGYA_PERSON_ID = "gigya_person_id"
GIGYA_KEYS = [GIGYA_JWT, GIGYA_LOGIN_TOKEN, GIGYA_PERSON_ID]

_LOGGER = logging.getLogger(__name__)


async def request(
    websession: aiohttp.ClientSession,
    method: str,
    url: str,
    data: Dict[str, Any],
    schema: Schema,
) -> models.GigyaResponse:
    """Send request to Gigya."""
    async with websession.request(method, url, data=data) as http_response:
        response_text = await http_response.text()
        # Disable logging on Gigya, to avoid unnecessary exposure.
        # _LOGGER.debug(
        #    "Received Gigya response %s on %s: %s",
        #    http_response.status,
        #    url,
        #    response_text,
        # )
        gigya_response: models.GigyaResponse = schema.loads(response_text)
        # Check for Gigya error
        gigya_response.raise_for_error_code()
        # Check for HTTP error
        http_response.raise_for_status()

        return gigya_response


async def login(
    websession: aiohttp.ClientSession,
    root_url: str,
    api_key: str,
    login_id: str,
    password: str,
) -> models.GigyaLoginResponse:
    """Send POST to /accounts.login."""
    return cast(
        models.GigyaLoginResponse,
        await request(
            websession,
            "POST",
            f"{root_url}/accounts.login",
            data={
                "ApiKey": api_key,
                "loginID": login_id,
                "password": password,
            },
            schema=schemas.GigyaLoginResponseSchema,
        ),
    )


async def get_account_info(
    websession: aiohttp.ClientSession,
    root_url: str,
    api_key: str,
    login_token: str,
) -> models.GigyaGetAccountInfoResponse:
    """Send POST to /accounts.getAccountInfo."""
    return cast(
        models.GigyaGetAccountInfoResponse,
        await request(
            websession,
            "POST",
            f"{root_url}/accounts.getAccountInfo",
            data={
                "ApiKey": api_key,
                "login_token": login_token,
            },
            schema=schemas.GigyaGetAccountInfoResponseSchema,
        ),
    )


async def get_jwt(
    websession: aiohttp.ClientSession,
    root_url: str,
    api_key: str,
    login_token: str,
) -> models.GigyaGetJWTResponse:
    """Send POST to /accounts.getJWT."""
    return cast(
        models.GigyaGetJWTResponse,
        await request(
            websession,
            "POST",
            f"{root_url}/accounts.getJWT",
            data={
                "ApiKey": api_key,
                "login_token": login_token,
                # gigyaDataCenter may be needed for future jwt validation
                "fields": "data.personId,data.gigyaDataCenter",
                "expiration": 900,
            },
            schema=schemas.GigyaGetJWTResponseSchema,
        ),
    )
