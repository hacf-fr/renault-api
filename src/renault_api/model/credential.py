"""Credential dataclass."""
import time
from dataclasses import dataclass

import jwt


@dataclass
class Credential:
    """Base credential that never expires."""

    value: str

    def has_expired(self) -> bool:
        """Check if Credential has expired."""
        return False


@dataclass
class JWTCredential(Credential):
    """JWT credential with expiry time."""

    expiry: float

    def __init__(self, value: str) -> None:
        """Initialise JWTCredential."""
        decoded_token = jwt.decode(value, options={"verify_signature": False})

        super().__init__(value=value)
        self.expiry = decoded_token["exp"]

    def has_expired(self) -> bool:
        """Check if JWT token has expired."""
        return self.expiry < time.time()
