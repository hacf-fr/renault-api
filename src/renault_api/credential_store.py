"""Kamereon client for interaction with Renault servers."""
import json
import os
from typing import Dict
from typing import List
from typing import Optional

import jwt

from renault_api.const import PERMANENT_KEYS
from renault_api.credential import Credential
from renault_api.credential import JWTCredential


class CredentialStore:
    """Credential store."""

    def __init__(self) -> None:
        """Initialise the credential store."""
        self._store: Dict[str, Credential] = {}

    def __getitem__(self, name: str) -> Credential:
        """Get a credential the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred
        raise KeyError(name)

    def get(self, name: str) -> Optional[Credential]:
        """Get a credential the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred
        return None

    def get_value(self, name: str) -> Optional[str]:
        """Get a credential value from the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred.value
        return None

    def __delitem__(self, name: str) -> None:
        """Remove a credential from the credential store."""
        del self._store[name]
        self._write()

    def __setitem__(self, name: str, value: Credential) -> None:
        """Add a credential to the credential store."""
        if not isinstance(name, str):  # pragma: no cover
            raise TypeError("`name` must be a string")
        if not isinstance(value, Credential):  # pragma: no cover
            raise TypeError("`value` must be a Credential")

        self._store[name] = value
        self._write()

    def __contains__(self, name: str) -> bool:
        """Check if a credential is in the credential store."""
        if name in self._store:
            cred = self._store[name]
            if not cred.has_expired():
                return True
        return False

    def _write(self) -> None:
        """Writes the content to fixed storage."""
        pass

    def clear(self) -> None:
        """Remove all non-permanent keys from credential store."""
        for key in list(self._store.keys()):
            if key not in PERMANENT_KEYS:
                del self._store[key]
        self._write()

    def clear_keys(self, to_delete: List[str]) -> None:
        """Remove specified keys from credential store."""
        for key in list(self._store.keys()):
            if key in to_delete:
                del self._store[key]
        self._write()


class CredentialEncoder(json.JSONEncoder):
    """Custom encoder for Credential class."""

    def default(self, obj: Credential) -> str:
        """Store the value."""
        return obj.value


class FileCredentialStore(CredentialStore):
    """Credential store with items stored in a file."""

    def __init__(self, store_location: str) -> None:
        """Initialise the credential store."""
        super().__init__()
        self._store_location = store_location
        self._read()

    def _read(self) -> None:
        """Read data from store location."""
        if not os.path.exists(self._store_location):
            return
        with open(self._store_location) as json_file:
            data = json.load(json_file)
            for key, value in data.items():
                if key == "gigya_jwt":
                    try:
                        self[key] = JWTCredential(value)
                    except jwt.ExpiredSignatureError:  # pragma: no cover
                        pass
                else:
                    self[key] = Credential(value)

    def _write(self) -> None:
        """Write data to store location."""
        dirname = os.path.dirname(self._store_location)
        if not os.path.isdir(dirname):
            os.makedirs(dirname)
        with open(self._store_location, "w") as json_file:
            json.dump(self._store, json_file, cls=CredentialEncoder)
