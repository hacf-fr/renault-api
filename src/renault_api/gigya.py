"""Gigya client for authentication."""
import logging
from typing import Any
from typing import cast
from typing import Dict

from aiohttp import ClientSession
from marshmallow.schema import Schema

from .const import CONF_GIGYA_APIKEY
from .const import CONF_GIGYA_URL
from renault_api.model import gigya as model

_LOGGER = logging.getLogger(__name__)


class Gigya:
    """Gigya client for authentication."""

    def __init__(
        self, websession: ClientSession, locale_details: Dict[str, str]
    ) -> None:
        """Initialise Gigya."""
        self._websession = websession

        self._api_key = locale_details[CONF_GIGYA_APIKEY]
        self._root_url = locale_details[CONF_GIGYA_URL]

    async def _post(
        self, path: str, data: Dict[str, Any], schema: Schema
    ) -> model.GigyaResponse:
        url = f"{self._root_url}{path}"
        async with self._websession.request("POST", url, data=data) as http_response:
            response_text = await http_response.text()
            _LOGGER.debug(
                "Received Gigya response %s on %s: %s",
                http_response.status,
                url,
                response_text,
            )
            gigya_response: model.GigyaResponse = schema.loads(response_text)
            # Check for Gigya error
            gigya_response.raise_for_error_code()
            # Check for HTTP error
            http_response.raise_for_status()

            return gigya_response

    async def login(self, login_id: str, password: str) -> model.GigyaLoginResponse:
        """POST to /accounts.login."""
        return cast(
            model.GigyaLoginResponse,
            await self._post(
                "/accounts.login",
                data={
                    "ApiKey": self._api_key,
                    "loginID": login_id,
                    "password": password,
                },
                schema=model.GigyaLoginResponseSchema,
            ),
        )

    async def get_account_info(
        self, login_token: str
    ) -> model.GigyaGetAccountInfoResponse:
        """POST to /accounts.getAccountInfo."""
        return cast(
            model.GigyaGetAccountInfoResponse,
            await self._post(
                "/accounts.getAccountInfo",
                data={"ApiKey": self._api_key, "login_token": login_token},
                schema=model.GigyaGetAccountInfoResponseSchema,
            ),
        )

    async def get_jwt(self, login_token: str) -> model.GigyaGetJWTResponse:
        """POST to /accounts.getAccountInfo."""
        return cast(
            model.GigyaGetJWTResponse,
            await self._post(
                "/accounts.getJWT",
                data={
                    "ApiKey": self._api_key,
                    "login_token": login_token,
                    # gigyaDataCenter may be needed for future jwt validation
                    "fields": "data.personId,data.gigyaDataCenter",
                    "expiration": 900,
                },
                schema=model.GigyaGetJWTResponseSchema,
            ),
        )
