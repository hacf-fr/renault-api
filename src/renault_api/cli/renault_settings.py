"""Singletons for the CLI."""
import os
from textwrap import TextWrapper
from typing import Any
from typing import Dict
from typing import Optional

import aiohttp
import click
from tabulate import tabulate

from renault_api.const import CONF_LOCALE
from renault_api.credential import Credential
from renault_api.credential_store import CredentialStore
from renault_api.helpers import get_api_keys

CONF_ACCOUNT_ID = "accound-id"
CONF_VIN = "vin"

CREDENTIAL_PATH = "~/.credentials/renault-api.json"


async def set_options(
    websession: aiohttp.ClientSession,
    ctx_data: Dict[str, Any],
    locale: Optional[str],
    account: Optional[str],
    vin: Optional[str],
) -> None:
    """Set configuration keys."""
    credential_store: CredentialStore = ctx_data["credential_store"]
    if locale:
        # Ensure API keys are available
        api_keys = await get_api_keys(locale, websession=websession)

        credential_store[CONF_LOCALE] = Credential(locale)
        for k, v in api_keys.items():
            credential_store[k] = Credential(v)

    if account:
        credential_store[CONF_ACCOUNT_ID] = Credential(account)
    if vin:
        credential_store[CONF_VIN] = Credential(vin)


def display_settings(ctx_data: Dict[str, Any]) -> None:
    """Get the current configuration keys."""
    credential_store: CredentialStore = ctx_data["credential_store"]
    wrapper = TextWrapper(width=80)
    items = list(
        [key, "\n".join(wrapper.wrap(credential_store.get_value(key) or "-"))]
        for key in credential_store._store.keys()
    )
    click.echo(tabulate(items, headers=["Key", "Value"]))


def reset() -> None:
    """Clear all credentials/settings from the credential store."""
    try:
        os.remove(os.path.expanduser(CREDENTIAL_PATH))
    except FileNotFoundError:
        pass
