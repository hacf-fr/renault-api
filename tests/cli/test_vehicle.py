"""Test cases for the __main__ module."""
import os
from locale import getdefaultlocale

from aioresponses import aioresponses
from click.testing import CliRunner
from tests import fixtures
from tests.const import TEST_ACCOUNT_ID
from tests.const import TEST_LOCALE
from tests.const import TEST_LOGIN_TOKEN
from tests.const import TEST_PASSWORD
from tests.const import TEST_PERSON_ID
from tests.const import TEST_USERNAME
from tests.const import TEST_VIN

from renault_api.cli import __main__
from renault_api.cli.renault_settings import CONF_ACCOUNT_ID
from renault_api.cli.renault_settings import CONF_VIN
from renault_api.cli.renault_settings import CREDENTIAL_PATH
from renault_api.const import CONF_LOCALE
from renault_api.credential import Credential
from renault_api.credential import JWTCredential
from renault_api.credential_store import FileCredentialStore
from renault_api.gigya import GIGYA_JWT
from renault_api.gigya import GIGYA_LOGIN_TOKEN
from renault_api.gigya import GIGYA_PERSON_ID

EXPECTED_BATTERY_STATUS = (
    "--------------------  -------------------------\n"
    "Battery level         50 %\n"
    "Last updated          2020-11-17 09:06:48\n"
    "Range estimate        128 km\n"
    "Plug state            PlugState.UNPLUGGED\n"
    "Charging state        ChargeState.NOT_IN_CHARGE\n"
    "Charge mode           always\n"
    "Total mileage         49114.27 km\n"
    "GPS Latitude          48.1234567\n"
    "GPS Longitude         11.1234567\n"
    "GPS last updated      2020-02-18 17:58:38\n"
    "HVAC status           off\n"
    "External temperature  8.0 Â°C\n"
    "--------------------  -------------------------\n"
)


def test_vehicle_status_prompt(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    fixtures.inject_gigya_all(mocked_responses)
    fixtures.inject_get_person(mocked_responses)

    # Injected for account selection
    fixtures.inject_get_vehicles(mocked_responses, "zoe_40.1")
    vehicle2_urlpath = f"accounts/account-id-2/vehicles?{fixtures.DEFAULT_QUERY_STRING}"
    fixtures.inject_data(
        mocked_responses,
        vehicle2_urlpath,
        "vehicles/zoe_40.1.json",
    )

    # Injected again for vehicle selection
    fixtures.inject_get_vehicles(mocked_responses, "zoe_40.1")
    fixtures.inject_vehicle_status(mocked_responses)

    result = cli_runner.invoke(
        __main__.main,
        "status",
        input=f"{TEST_LOCALE}\nN\n{TEST_USERNAME}\n{TEST_PASSWORD}\n1\ny\n1\ny\n",
    )
    assert result.exit_code == 0, result.exception

    expected_output = (
        f"Please select a locale [{getdefaultlocale()[0]}]: {TEST_LOCALE}\n"
        "Do you want to save the locale to the credential store? [y/N]: N\n"
        "\n"
        f"User: {TEST_USERNAME}\n"
        "Password: \n"
        "\n"
        "    ID            Type         Vehicles\n"
        "--  ------------  ---------  ----------\n"
        " 1  account-id-1  MYRENAULT           1\n"
        " 2  account-id-2  SFDC                1\n"
        "\n"
        "Please select account [1]: 1\n"
        "Do you want to save the account ID to the credential store? [y/N]: y\n"
        "\n"
        "    Vin                Registration    Brand    Model\n"
        "--  -----------------  --------------  -------  -------\n"
        " 1  VF1AAAAA555777999  REG-NUMBER      RENAULT  ZOE\n"
        "\n"
        "Please select vehicle [1]: 1\n"
        "Do you want to save the VIN to the credential store? [y/N]: y\n"
        "\n"
        f"{EXPECTED_BATTERY_STATUS}"
    )
    assert expected_output == result.output


def test_vehicle_status_store(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    credential_store[CONF_LOCALE] = Credential(TEST_LOCALE)
    credential_store[CONF_ACCOUNT_ID] = Credential(TEST_ACCOUNT_ID)
    credential_store[CONF_VIN] = Credential(TEST_VIN)
    credential_store[GIGYA_LOGIN_TOKEN] = Credential(TEST_LOGIN_TOKEN)
    credential_store[GIGYA_PERSON_ID] = Credential(TEST_PERSON_ID)
    credential_store[GIGYA_JWT] = JWTCredential(fixtures.get_jwt())

    fixtures.inject_vehicle_status(mocked_responses)

    result = cli_runner.invoke(__main__.main, "status")
    assert result.exit_code == 0, result.exception

    assert EXPECTED_BATTERY_STATUS == result.output


def test_vehicle_status_no_prompt(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    credential_store[CONF_LOCALE] = Credential(TEST_LOCALE)
    credential_store[GIGYA_LOGIN_TOKEN] = Credential(TEST_LOGIN_TOKEN)
    credential_store[GIGYA_PERSON_ID] = Credential(TEST_PERSON_ID)
    credential_store[GIGYA_JWT] = JWTCredential(fixtures.get_jwt())

    fixtures.inject_vehicle_status(mocked_responses)

    result = cli_runner.invoke(
        __main__.main, f"--account {TEST_ACCOUNT_ID} --vin {TEST_VIN} status"
    )
    assert result.exit_code == 0, result.exception

    assert EXPECTED_BATTERY_STATUS == result.output
