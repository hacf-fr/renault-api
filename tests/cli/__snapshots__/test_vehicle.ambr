# serializer version: 1
# name: test_http_get
  '''
  {'data': {'type': 'Car', 'id': 'VF1AAAAA555777999', 'attributes': {'calendar': {'monday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'tuesday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'wednesday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'thursday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'friday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'saturday': [{'startTime': '2300', 'duration': 480, 'activationState': False}], 'sunday': [{'startTime': '2300', 'duration': 480, 'activationState': False}]}}}}
  
  '''
# ---
# name: test_http_get_list
  '''
  {'data': [{'notificationId': 'ffcb0310-503f-4bc3-9056-e9d051a089c6', 'notifDate': '2022-02-01T19:01:51.622', 'vin': '*PRIVATE*', 'personId': '*PRIVATE*', 'kmrUserId': '*PRIVATE*', 'actionType': 'COMMAND_RESPONSE', 'commandResponse': {'status': 'CREATED'}, 'commandType': 'SRP_SETS'}, {'notificationId': 'ffcb0310-503f-4bc3-9056-e9d051a089c6', 'notifDate': '2022-02-01T19:01:51.623', 'vin': '*PRIVATE*', 'personId': '*PRIVATE*', 'kmrUserId': '*PRIVATE*', 'actionType': 'SRP_SALT_REQUEST', 'srpResponse': {'status': 'OK', 'loginB': '*PRIVATE*', 'loginSalt': '*PRIVATE*'}}]}
  
  '''
# ---
# name: test_http_post
  '''
  {'data': {'type': 'ChargeSchedule', 'id': 'guid', 'attributes': {'schedules': [{'id': 1, 'activated': True, 'tuesday': {'startTime': 'T04:30Z', 'duration': 420}, 'wednesday': {'startTime': 'T22:30Z', 'duration': 420}, 'thursday': {'startTime': 'T22:00Z', 'duration': 420}, 'friday': {'startTime': 'T23:30Z', 'duration': 480}, 'saturday': {'startTime': 'T18:30Z', 'duration': 120}, 'sunday': {'startTime': 'T12:45Z', 'duration': 45}}]}}}
  
  '''
# ---
# name: test_http_post.1
  dict({
    'data': dict({
      'attributes': dict({
        'schedules': list([
        ]),
      }),
      'type': 'ChargeSchedule',
    }),
  })
# ---
# name: test_http_post_file
  '''
  {'data': {'type': 'ChargeSchedule', 'id': 'guid', 'attributes': {'schedules': [{'id': 1, 'activated': True, 'tuesday': {'startTime': 'T04:30Z', 'duration': 420}, 'wednesday': {'startTime': 'T22:30Z', 'duration': 420}, 'thursday': {'startTime': 'T22:00Z', 'duration': 420}, 'friday': {'startTime': 'T23:30Z', 'duration': 480}, 'saturday': {'startTime': 'T18:30Z', 'duration': 120}, 'sunday': {'startTime': 'T12:45Z', 'duration': 45}}]}}}
  
  '''
# ---
# name: test_http_post_file.1
  dict({
    'data': dict({
      'attributes': dict({
        'schedules': list([
        ]),
      }),
      'type': 'ChargeSchedule',
    }),
  })
# ---
# name: test_vehicle_contracts
  '''
  Type                            Code                  Description                       Start       End         Status
  ------------------------------  --------------------  --------------------------------  ----------  ----------  ------------------
  WARRANTY_MAINTENANCE_CONTRACTS  40                    CONTRAT LOSANGE                   2018-04-04  2022-04-03  Actif
  CONNECTED_SERVICES              ZECONNECTP            My Z.E. Connect en série 36 mois  2018-08-23  2021-08-23  Actif
  CONNECTED_SERVICES              GBA                   Battery Services                  2018-03-23              Echec d’activation
  WARRANTY                        ManufacturerWarranty  Garantie fabricant                            2020-04-03  Expiré
  WARRANTY                        PaintingWarranty      Garantie peinture                             2021-04-03  Actif
  WARRANTY                        CorrosionWarranty     Garantie corrosion                            2030-04-03  Actif
  WARRANTY                        GMPeWarranty          Garantie GMPe                                 2020-04-03  Expiré
  WARRANTY                        AssistanceWarranty    Garantie assistance                           2020-04-03  Expiré
  
  '''
# ---
# name: test_vehicle_details
  '''
  Registration    Brand    Model    VIN
  --------------  -------  -------  -----------------
  REG-NUMBER      RENAULT  ZOE      VF1AAAAA555777999
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/captur_ii.1.json]
  '''
  --------------------  ----------------------
  Total mileage         5566.78 km
  Fuel autonomy         35.0 km
  Fuel quantity         3.0 L
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Lock status           locked
  Lock last updated     2022-02-02 14:51:13
  Engine state          Stopped, ready for RES
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  ----------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/captur_ii.2.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         5566.78 km
  Fuel autonomy         35.0 km
  Fuel quantity         3.0 L
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Lock status           locked
  Lock last updated     2022-02-02 14:51:13
  Engine state          Stopped, ready for RES
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/clio_v.1.json]
  '''
  --------------------  ----------------------
  Total mileage         49114.27 km
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Lock status           locked
  Lock last updated     2022-02-02 14:51:13
  Engine state          Stopped, ready for RES
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  ----------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/duster.1.json]
  '''
  --------------------  ----------------------
  Total mileage         49114.27 km
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Lock status           locked
  Lock last updated     2022-02-02 14:51:13
  Engine state          Stopped, ready for RES
  HVAC status           off
  External temperature  8.0 °C
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  ----------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/megane_e-tech.1.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Engine state          Stopped, ready for RES
  HVAC status           off
  External temperature  8.0 °C
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/megane_e-tech.2.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Engine state          Stopped, ready for RES
  HVAC status           off
  External temperature  8.0 °C
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/renault_5.1.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Lock status           locked
  Lock last updated     2022-02-02 14:51:13
  Engine state          Stopped, ready for RES
  HVAC status           off
  External temperature  8.0 °C
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/spring.1.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Lock status           locked
  Lock last updated     2022-02-02 14:51:13
  Engine state          Stopped, ready for RES
  HVAC status           off
  External temperature  8.0 °C
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/twingo_ze.1.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Lock status           locked
  Lock last updated     2022-02-02 14:51:13
  Engine state          Stopped, ready for RES
  HVAC status           on
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/zoe_40.1.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  HVAC status           off
  External temperature  8.0 °C
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/zoe_40.2.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  HVAC status           off
  External temperature  8.0 °C
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status[tests/fixtures/kamereon/vehicles/zoe_50.1.json]
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         5785.75 km
  Fuel autonomy         0.0 km
  Fuel quantity         0.0 L
  GPS Latitude          48.1234567
  GPS Longitude         11.1234567
  GPS last updated      2020-02-18 17:58:38
  Lock status           locked
  Lock last updated     2022-02-02 14:51:13
  Engine state          Stopped, ready for RES
  HVAC status           on
  Front left pressure   2460 bar
  Front right pressure  2730 bar
  Rear left pressure    2790 bar
  Rear right pressure   2790 bar
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status_json
  '''
  {"battery-status": {"timestamp": "2020-11-17T09:06:48+01:00", "batteryLevel": 50, "batteryAutonomy": 128, "batteryCapacity": 0, "batteryAvailableEnergy": 0, "plugStatus": 0, "chargingStatus": -1.0}, "charge-mode": {"chargeMode": "always"}, "cockpit": {"totalMileage": 49114.27}, "hvac-status": {"externalTemperature": 8.0, "hvacStatus": "off"}}
  
  '''
# ---
# name: test_vehicle_status_no_prompt
  '''
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  HVAC status           off
  External temperature  8.0 °C
  --------------------  -------------------------
  
  '''
# ---
# name: test_vehicle_status_prompt
  '''
  Please select a locale [fr_CA]: fr_FR
  Do you want to save the locale to the credential store? [y/N]: N
  
  User: test@example.com
  Password: 
  
      ID            Type         Vehicles
  --  ------------  ---------  ----------
   1  account-id-1  MYRENAULT           1
   2  account-id-2  SFDC                1
  
  Please select account [1]: 1
  Do you want to save the account ID to the credential store? [y/N]: y
  
      Vin                Registration    Brand    Model
  --  -----------------  --------------  -------  -------
   1  VF1AAAAA555777999  REG-NUMBER      RENAULT  ZOE
  
  Please select vehicle [1]: 1
  Do you want to save the VIN to the credential store? [y/N]: y
  
  --------------------  -------------------------
  Battery level         50 %
  Last updated          2020-11-17 09:06:48
  Range estimate        128 km
  Plug state            PlugState.UNPLUGGED
  Charging state        ChargeState.NOT_IN_CHARGE
  Charge mode           always
  Total mileage         49114.27 km
  HVAC status           off
  External temperature  8.0 °C
  --------------------  -------------------------
  
  '''
# ---
